### DISCUSSION TITLE
Preserving upstream README, CHANGELOG, etc. files in a fork using Actions.

### DISCUSSION THREAD

**Discussion by erictheise** [community/community#123667] [Other Feature Feedback, Questions, & Ideas]
### Select Topic Area

Question

### Body

I'm working on a forked project. While we may occasionally contribute to the upstream project the main purpose of our fork is adding functionality that is not appropriate for the upstream one. I want to use GitHub Actions to periodically pull in upstream changes and run tests in an attempt to stay in sync. Common use case.

As part of this I'd like to keep standard upstream files but also have our own that take precedence. What I mean by that is the upstream `README.md`, `CHANGELOG.md`, `LICENCE`, `TODO.md`, etc. should be available in our repo, say in an `/upstream` subdirectory, and that ours should be at the root level of the project.

What are the best ways of accomplishing this kind of renaming/moving/preserving upstream files? Thanks in advance.

**Comment by spenserblack** [comment]
An interesting one for sure!

To run periodically, you probably want to [run your action on a schedule](https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#schedule).

But, as far as *what* to actually run on the schedule, here are some git commands that might help you:
- `git merge --no-ff --no-commit SOURCE`: Start a merge, but make sure that the merge is paused
  - `--no-ff` ensures that the merge isn't fast-forwarded. It will stop git from placing the new commits on top of your current commits *without a merge commit.*
  - `--no-commit` ensures that git doesn't automatically try to create a commit when merging
- `git checkout [--ours|--theirs] -- FILEPATH`: this allows you to checkout a file from a "side" of the pending commit. It's probably easiest to understand by playing around with this.

I'd recommend writing a shell script that:
1. Fetches the upstream
2. Starts a merge
3. Runs `git checkout --theirs` (or `--ours`) on the appropriate files.
4. Commits the merge

This way you can easily play around with this script locally, and when you're satisfied with its behavior you can write an action to run this script on a schedule.

**Comment by erictheise** [comment]
Belated thanks for your suggestions, @spenserblack, they were very helpful. I did create a bash script that did most of the work and I've slowly been shifting commands from that script to the workflow when it seems appropriate.

I mentioned one goal as wanting to preserve the upstream documentation in its own subdirectory and `git show` turned out to be a good approach; here it's used to grab any markdown files and the `LICENSE` file.

```
mkdir upstream
for i in $(git ls-tree upstream/master --name-only); do
  if [[ $i == *.md ]];then
    git show upstream/master:$i > upstream/$i
  fi
done
git show upstream/master:LICENSE > upstream/LICENSE
git add upstream
if [ -n "$(git status --porcelain)" ]; then
  git commit -m 'Preserving upstream markdown & text files.'
fi
```
